server/task/src/main/java/org/apache/james/task/MemoryTaskManager.java

```diff
@@ -128,32 +86,45 @@ public TaskExecutionDetails getExecutionDetails(TaskId id) {
 
     @Override
     public List<TaskExecutionDetails> list(Status status) {
-        return idToExecutionDetails.values()
+        return ImmutableList.copyOf(tasksFiltered(status).values());
+    }
+
+    public Set<TaskId> listIds(Status status) {
+        return tasksFiltered(status).keySet();
+    }
+
+    public Map<TaskId, TaskExecutionDetails> tasksFiltered(Status status) {
+        return idToExecutionDetails.entrySet()
             .stream()
-            .filter(details -> details.getStatus().equals(status))
-            .collect(Guavate.toImmutableList());
+            .filter(details -> details.getValue().getStatus().equals(status))
+            .collect(Guavate.entriesToImmutableMap());
     }
 
     @Override
     public void cancel(TaskId id) {
-        Optional.ofNullable(idToFuture.get(id))
-            .ifPresent(future -> {
-                TaskExecutionDetails executionDetails = idToExecutionDetails.get(id);
-                idToExecutionDetails.put(id, executionDetails.cancel());
-                future.cancel(INTERRUPT_IF_RUNNING);
-                idToFuture.remove(id);
-            });
+        TaskExecutionDetails details = getExecutionDetails(id);
+        if (details.getStatus().equals(Status.IN_PROGRESS) || details.getStatus().equals(Status.WAITING)) {
```

Comment:
you already check details status in the worker, why here also?

Reply:
adding a method `isInStatus(Status...status)` can helps to save a long line

Reply:
you are right it seems unnecessary here.